SPEC-1690 Add average RTT tracking to server monitoring pseudocode

ShaneHarvey opened the pull request:
Jira: [SPEC-1690](https://jira.mongodb.org/browse/SPEC-1690)

divjotarora commented on source/server-discovery-and-monitoring/server-monitoring.rst:
```diff
@@ -682,27 +686,29 @@ The event API here is assumed to be like the standard `Python Event
+                response = read next isMaster exhaust response
+            elif previousDescription.topologyVersion:
                 # Initiate streaming isMaster
+                rttMonitor.start()
```
nit: In the changes for the RTT thread description below, this method is called `run`.

ShaneHarvey replied:
Fixed.

ShaneHarvey commented:
> The current spec wording for the Measuring RTT section says that clients can use a dedicated RTT connection for all servers, even those that don't support the streamable protocol. I found it easier to implement that way in Go, but the pseudocode seems to only use the RTT thread for streaming. Was this intentional? 

Yes it was intentional. I modeled this after the Python implementation which only starts the RttMonitor when it is discovered that the server supports streaming. Either way is fine so I'm open to changing the pseudocode to always start the RttMonitor.

> Also, the RTT monitor should be stopped if the server replies without a topologyVersion. 

Do we need to? It seems fine to continue using the RttMonitor after a downgrade because it's valid to use the RttMonitor on all server versions. 

> As an aside, the pseudocode doesn't handle the case of a server omitting topologyVersion when one was previously included. The spec says that this isn't an expected state and the driver should restart monitoring on a new connection if it happens.

I think we should remove the line that talks about this case because it is [guaranteed by the server not to happen](https://github.com/mongodb/specifications/blob/master/source/server-discovery-and-monitoring/server-monitoring.rst#awaitable-ismaster-protocol):
```
A server that implements the new protocol follows these rules:

- Always include the server's topologyVersion in isMaster and State Change
  Error replies.
```

divjotarora commented:
> Do we need to? It seems fine to continue using the RttMonitor after a downgrade because it's valid to use the RttMonitor on all server versions.

I agree it's not incorrect to keep it running. We'll just end up taking more samples than previous releases because both the RTTMonitor and the non-streaming/awaitable heartbeat paths add samples. I suggested stopping it on downgrades because my understanding is that starting it only when we detect that the server supports streaming is useful for making sure we don't use multiple dedicated connections if they're not needed, but this wouldn't be true for downgrades.

> I think we should remove the line that talks about this case because it is guaranteed by the server not to happen

Agreed. If the server is already asserting that a 4.4+ node can't return an `{ok: 1}` response without `topologyVersion`, we shouldn't be checking driver-side. @jyemin @mbroadst Thoughts on removing that line?

kevinAlbs commented on source/server-discovery-and-monitoring/server-monitoring.rst:
```diff
@@ -682,27 +686,29 @@ The event API here is assumed to be like the standard `Python Event
+                response = call {isMaster: 1}
+                rttMonitor.addSample(time()-start)
+
+            return ServerDescription(response, rtt=rttMonitor.average())
```
Apologies for barging, but this seemed like an apt PR to bring this up. I interpret this to mean that we should only be using the RTT from the server monitor. Is it permitted for application handshakes to update RTT?

ShaneHarvey replied:
I think that's allowed. In fact, if an application handshake can transition a server from Unknown to Known then it might be required to update the RTT as well (to avoid having a 0 value for average RTT). FYI Python/Java do not update the topology from application handshakes.

kevinAlbs replied:
In that case, I may diverge a little from pseudocode to implement RTT monitoring. The server description on the topology may have an existing running average updated by application handshakes, so I cannot overwrite it.

A related question came to mind. Are we explicitly deciding not to update the topology description when RTT receives a response? There is the note to not mark servers unknown on RTT failure:
> Errors encountered when running a "isMaster" command MUST NOT update the topology.

If it's safe to do so, I could just update the RTT of in the topology description as soon as the RTT monitor thread receives an ismaster reply (including the handshake).

ShaneHarvey replied:
> Are we explicitly deciding not to update the topology description when the RTT monitor receives a response?

Yes. We decided not to update the topology to avoid churn and to avoid races (incorrectly going from Unknown to known or vice versa). The RttThread can't just publish new ServerDescriptions the same way the Monitor does without adding races. There would need to be a new method that simply updates a server's average rtt.

Could you describe how your idea would work in more detail?

kevinAlbs replied:
Ah gotcha. I can simply update the server's average RTT then. I can take a stab at noting that in the "Measuring RTT" section if that seems worth calling out.

What I have working now is an RTT monitor thread that behaves close to the pseudocode, except shared ServerDescription in the TopologyDescription is responsible for the RTT tracking. When an RTT ismaster or polling ismaster returns, it will add to the shared ServerDescription's RTT tracking. 

Updating just the RTT in a ServerDescription seems natural in libmongoc. In the current behavior, ServerDescription's are not immutable and are responsible for updating the RTT. When a server monitor finishes a polling ismaster, the topology description is fed the BSON reply + RTT or an error. [This is the code path](https://github.com/mongodb/mongo-c-driver/blob/9f43168cfa94313d9e7aeffa75d95f76fb25db50/src/libmongoc/src/mongoc/mongoc-topology.c#L90-L91) where either a server monitor and application handshake updates the topology description after an ismaster attempt. The optional RTT there is the RTT for that single ismaster. Though, FWIW I'm considering making a ServerDescription immutable in libmongoc. It'd be easier to reason about, and is more in line with the SDAM spec's recommendations. But I think the logic for updating just the RTT could be the same.

kevinAlbs replied:
Made a note in https://github.com/mongodb/specifications/pull/810

ShaneHarvey replied:
Thanks, I think that sounds reasonable given that C's ServerDescription is mutable.

ShaneHarvey commented:
Rebased to resolve the merge confilcts.

divjotarora commented on source/server-discovery-and-monitoring/server-monitoring.rst:
```diff
@@ -742,6 +752,66 @@ The event API here is assumed to be like the standard `Python Event
+            try:
+                rtt = pingServer()
+                addSample(rtt)
+            except (NetworkError, CommandError) as exc:
```
Do we need to reset the average RTT here or is it enough for the server monitor to reset it when it errors?

ShaneHarvey replied:
I don't think we should. If this thread calls reset() then it might lead to this race:
```
    def run():
....
          else:
              # The server does not support topologyVersion.
              set connection timeout to connectTimeoutMS
              start = time()
              response = call {isMaster: 1}
              rttMonitor.addSample(time()-start)
            # <--- RttMonitor calls reset() here. rttMonitor.average() is now 0.
            return ServerDescription(response, rtt=rttMonitor.average())
```

divjotarora replied:
Agreed. I asked because I talked to @jyemin about the Java code and he had structured it in a way where both the server and RTT monitors reset the RTT, so we weren't sure which one(s) should.

mbroadst commented:
@divjotarora yes, let's go ahead and remove that line from the spec. It's quite difficult to test, and the server guarantees the behavior.

@ShaneHarvey @divjotarora Regarding the RTT tracking: I think it was a mistake of convenience that we every tied RTT tracking to server heartbeats, we happened to be doing something every 10s that was close to an "immediate" command, so why not use it to measure round trip time? It's a side effect.

What if instead we _always_ require an "RTT Monitor" and treat it as the source of truth for these calculations? I think that makes the design easier to reason about, and makes it very explicit where and how the RTT is measured and calculated. We could leave in an acceptable deviation for drivers who wish to do something more complicated (switch between RTT tracker and `isMaster` calls, do both at the same time, etc). Thoughts?

ShaneHarvey commented:
> What if instead we always require an "RTT Monitor" and treat it as the source of truth for these calculations? I think that makes the design easier to reason about, and makes it very explicit where and how the RTT is measured and calculated. We could leave in an acceptable deviation for drivers who wish to do something more complicated (switch between RTT tracker and isMaster calls, do both at the same time, etc). Thoughts?

I've removed the missing topologyVersion line and updated the code to always start the RttMonitor. However, I still think that the RttMonitor cannot reset the average RTT to 0 without introducing the race mentioned above ([0 RTT with a known ServerDescription](https://github.com/mongodb/specifications/pull/807#discussion_r432687114)).

divjotarora commented:
I agree that we should always require the RTT monitor and that the RTT monitor should not call reset. If there's a long-lived issue on the server, the server monitor will reset. If it's transient and only hit by the RTT monitor once, resetting doesn't really benefit us anyway.

ShaneHarvey commented:
> I agree that we should always require the RTT monitor and that the RTT monitor should not call reset. If there's a long-lived issue on the server, the server monitor will reset. If it's transient and only hit by the RTT monitor once, resetting doesn't really benefit us anyway.

So you propose that we never reset the RTT and remove [this requirement from the server selection spec](https://github.com/mongodb/specifications/blob/master/source/server-selection/server-selection.rst#calculation-of-average-round-trip-times)?:
> An Unknown server has no average RTT. When a server becomes unavailable, its average RTT MUST be cleared. Clients MAY implement this idiomatically (e.g nil, -1, etc.)

divjotarora commented:
No, I was just agreeing that the RTT monitor should not reset the value because it makes things harder to reason about and also causes the race you mentioned. I think it's fine for the server monitor  to reset the average RTT, though.

On an unrelated note, is there any way we can test that drivers use the initial connection handshake from the ServerMonitor to update the average RTT?

ShaneHarvey commented:
Anything else @mbroadst?

jyemin commented on source/server-discovery-and-monitoring/server-monitoring.rst:
```````````diff
@@ -742,6 +747,66 @@ The event API here is assumed to be like the standard `Python Event
+RTT thread
+``````````
+
+The requirements in the `Measuring RTT`_ section can be statisfied with an
```````````
statisfied => satisfied

ShaneHarvey replied:
Done.

DmitryLukyanov commented on source/server-discovery-and-monitoring/server-monitoring.rst:
```diff
@@ -742,6 +747,66 @@ The event API here is assumed to be like the standard `Python Event
+            try:
+                rtt = pingServer()
+                addSample(rtt)
+            except (NetworkError, CommandError) as exc:
```
as I understand any other exceptions except `NetworkError, CommandError` is not possible here. if so, will it be more accurate to have only `except (Exception)` line here?

ShaneHarvey replied:
Good point, I updated to catch all exceptions.
